druid/src/widget/image.rs

```diff
@@ -190,19 +190,22 @@ impl<T: Data> Widget<T> for Image {
         bc.debug_check("Image");
 
         // If either the width or height is constrained calculate a value so that the image fits
-        // in the size exactly. If it is unconstrained by both width and height take the size of
-        // the image.
+        // in the size exactly. If it is unconstrained by both width and height then use the fill
+        // strategy to determine the size.
         let max = bc.max();
         let image_size = self.image_data.size();
-        let size = if bc.is_width_bounded() && !bc.is_height_bounded() {
-            let ratio = max.width / image_size.width;
-            Size::new(max.width, ratio * image_size.height)
-        } else if bc.is_height_bounded() && !bc.is_width_bounded() {
-            let ratio = max.height / image_size.height;
-            Size::new(ratio * image_size.width, max.height)
-        } else {
-            bc.constrain(self.image_data.size())
-        };
+        let affine = self.fill.affine_to_fill(max, image_size).as_coeffs();
+        // The first and forth elements of the affine are the x and y scale factor.
+        // So just multiply them by the original size to get the ideal area based on `self.fill`.
+        let mut width = affine[0] * image_size.width;
+        let mut height = affine[3] * image_size.height;
+        if bc.is_width_bounded() {
+            width = width.min(max.width);
+        }
+        if bc.is_height_bounded() {
+            height = height.min(max.height);
+        }
```

Comment:
I think the strategy to use `affine_to_fill` is probably good. However this doesnt take into account when both sides are unbounded.

In my if else chain the first 2 should remain the same, since this takes into account being inside unbounded parents. This code should only apply if its bounded on both sides, since `affine_to_fill` is not made for unbounded constraints. In the last case where all sides are unbounded it should use the `bc.constrain(self.image_data.size())` code (Which could be simplified to `self.image_data.size()` since bc is unbounded). 

The reason the code is the way it is now, is so that we dont return an infinite size when unbounded in either direction. Since your code doesnt take into account being unbounded on either side we still need these safeguards using a different implementation.

Reply:
> Since your code doesnt take into account being unbounded on either side we still need these safeguards using a different implementation.

But this code never returns a unbounded result. and some of the fill strategies call for more than the image. So letting `affine_to_fill` determine the result dose all the work you are asking for when you want it and other things for the other fill strategies.

`self.fill.affine_to_fill(max, image_size)` takes the max size the widget may be to not change the window size, this gives it a `max` to work with. then it also knows the original image size, `image_size`. so for `no scale` or `scale down` the max it asks for the smallest of original image or the space available.

`affine_to_fill` is doing the guarding its self in the `non-constrained` case and the `fully constrained` case

Reply:
I was looking at this code and realised I can simplify it without changing the behaviour of the demo at all. Also by having less branching it makes it simpler for the cpu so might be a tiny tiny bit faster.

I have added some comments that might make it easer to under stand as well.

Reply:
>But this code never returns a unbounded result

It does, when the constraints are unbounded the resulting size of the image could be infinitely big. Since there are no checks for this this infinity will traverse it's way through the calculations into the final size.

> or the space available.

that's the problem, the space available could be infinite in either direction.